File: poker-client/src/main/java/pokergame/client/utils/EventBus.java
package pokergame.client.utils;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;

public class EventBus {
    private static final Map<Class<?>, List<Consumer<Object>>> listeners = new ConcurrentHashMap<>();

    /**
     * Subscribes a handler callback to a specific class type payload.
     */
    @SuppressWarnings("unchecked")
    public static <T> void subscribe(Class<T> eventType, Consumer<T> handler) {
        listeners.computeIfAbsent(eventType, k -> new ArrayList<>())
                .add((Consumer<Object>) handler);
    }

    /**
     * Unsubscribes a specific handler to prevent memory leaks and duplicate UI updates.
     */
    public static <T> void unsubscribe(Class<T> eventType, Consumer<T> handler) {
        List<Consumer<Object>> handlers = listeners.get(eventType);
        if (handlers != null) {
            handlers.remove(handler);
        }
    }

    /**
     * Publishes an event instance to all registered handlers matching its exact class type.
     */
    public static void publish(Object event) {
        if (event == null) return;
        Class<?> eventType = event.getClass();
        List<Consumer<Object>> handlers = listeners.get(eventType);

        if (handlers != null) {
            // Operate on a copy to prevent ConcurrentModificationException during scene changes
            for (Consumer<Object> handler : new ArrayList<>(handlers)) {
                try {
                    handler.accept(event);
                } catch (Exception e) {
                    System.err.println("[EventBus Error] Subscriber handling failed: " + e.getMessage());
                }
            }
        }
    }

    /**
     * Optional utility to clear subscribers when changing scenes or tearing down tables
     */
    public static void clear() {
        listeners.clear();
    }
}

File: poker-client-unity/Assets/Core/Utils/EventBus.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using UnityEngine;

namespace PokerGame.Client.Utils
{
    public static class EventBus
    {
        private static readonly ConcurrentDictionary<Type, List<Delegate>> Listeners = new();

        public static void Subscribe<T>(Action<T> handler)
        {
            if (handler == null) return;
            var type = typeof(T);
            Listeners.AddOrUpdate(type,
                _ => new List<Delegate> { handler },
                (_, list) =>
                {
                    lock (list) { list.Add(handler); }
                    return list;
                });
        }

        public static void Unsubscribe<T>(Action<T> handler)
        {
            if (handler == null) return;
            var type = typeof(T);
            if (Listeners.TryGetValue(type, out var list))
            {
                lock (list) { list.Remove(handler); }
            }
        }

        public static void Publish<T>(T eventMessage)
        {
            if (eventMessage == null) return;
            var type = typeof(T);
            if (Listeners.TryGetValue(type, out var list))
            {
                List<Delegate> copy;
                lock (list) { copy = new List<Delegate>(list); }

                foreach (var listener in copy)
                {
                    try
                    {
                        ((Action<T>)listener)?.Invoke(eventMessage);
                    }
                    catch (Exception ex)
                    {
                        Debug.LogError($"[EventBus Error] Subscriber handling failed: {ex.Message}");
                    }
                }
            }
        }

        public static void Clear() => Listeners.Clear();
    }
}